src/ipc/ipc-zephyr.c

```diff
@@ -79,6 +83,62 @@ static bool message_handler(const struct device *dev, void *arg, uint32_t data,
 	return false;
 }
 
+#ifdef CONFIG_PM_DEVICE
+/**
+ * @brief IPC device suspend handler callback function.
+ * Checks whether device power state should be actually changed.
+ *
+ * @param dev IPC device.
+ * @param arg IPC struct pointer.
+ */
+static int ipc_device_suspend_handler(const struct device *dev, void *arg)
+{
+	struct ipc *ipc = (struct ipc *)arg;
+
+	// we are not entering D3 - return error code bad message
+	if (!(ipc->task_mask & IPC_TASK_POWERDOWN))
+		return -EBADMSG;
+
+	// we are not entering D3 - power state change is not caused by IPC message
+	// from host that would disable the lock.
+	if (pm_policy_state_lock_is_active(PM_STATE_SOFT_OFF, PM_ALL_SUBSTATES))
+		return -EACCES;
+
+	return 0;
+}
+
+/**
+ * @brief IPC device resume handler callback function.
+ * Resets IPC control after context restore.
+ *
+ * @param dev IPC device.
+ * @param arg IPC struct pointer.
+ */
+static int ipc_device_resume_handler(const struct device *dev, void *arg)
+{
+	struct ipc *ipc = (struct ipc *)arg;
+
+	ipc_set_drvdata(ipc, NULL);
+	ipc->task_mask = 0;
+	ipc->pm_prepare_D3 = false;
+
+	// re-enable D3 power state lock
+	pm_policy_state_lock_get(PM_STATE_SOFT_OFF, PM_ALL_SUBSTATES);
+
+	/* attach handlers */
+	intel_adsp_ipc_set_message_handler(INTEL_ADSP_IPC_HOST_DEV, message_handler, ipc);
```

Comment:
Not related to this patch, but realized this interface is not really well named, this should be used for all SOF platforms, not just Intel DSPs, so "intel_adsp_" prefix is not good here. But alas this is not added by this PR. FYI to @juimonen